code/centauri/light-clients/ics10-grandpa/src/client_def.rs

```diff
@@ -94,24 +97,73 @@ where
 				.map_err(Error::GrandpaPrimitives)?;
 			},
 			ClientMessage::Misbehaviour(misbehavior) => {
-				// first off is the number of equivocations >= 1/3?
-				if misbehavior.equivocations.len() < (client_state.current_authorities.len() / 3) {
+				let first_proof = misbehavior.first_finality_proof;
+				let second_proof = misbehavior.second_finality_proof;
+
+				let first_header = first_proof.unknown_headers.first().ok_or_else(|| {
+					Error::Custom("No headers in first finality proof".to_string())
+				})?;
+
+				let second_header = second_proof.unknown_headers.first().ok_or_else(|| {
+					Error::Custom("No headers in second finality proof".to_string())
+				})?;
+
+				let first_parent = first_header.parent_hash;
+				let second_parent = second_header.parent_hash;
+
+				let check_canonicity = |chain: &[RelayChainHeader], init: H256| {
+					let mut last_hash = first_parent;
+					for header in chain {
+						if header.hash() != last_hash {
+							return Err(Error::Custom("Chain is not canonical".to_string()))
+						}
+						last_hash = header.parent_hash;
+					}
+					Ok(())
+				};
+
+				check_canonicity(&first_proof.unknown_headers, first_parent)?;
+				check_canonicity(&second_proof.unknown_headers, second_parent)?;
```

Comment:
[nit] an alternative is to accept the two proofs in the `check_canonicity`

Reply:
`check_canonicity` could accept a slice of proofs for instance instead of calling it twice. But it's a minor thing, you choose :)

Reply:
That might make the function less succint.